### ISSUE TITLE
_minimize_writes re-creating unmodified files

### ISSUE THREAD

**Issue by lnussel** [rpm-software-management/rpm#1881]
openSUSE Tumbleweed with rpm 4.17.0 running with `%_minimize_wites 1`. I noticed that some package updates lead to inaccessible dbus services. Rpm intends to only touch unchanged files but ends up replacing them. Eg. (--force doesn't matter, happens on regular updates too)

```
linux:~ # rpm -Uvvh systemd-container-250+git20220110.bf732cf8c5-0.x86_64.rpm --force --fsmdebug 2>log
Verifying...                          ################################# [100%]
Preparing...                          ################################# [100%]
Updating / installing...
   1:systemd-container-250+git20220110################################# [ 50%]
Cleaning up / removing...
   2:systemd-container-250+git20220110################################# [100%]
linux:~ # grep machine1.conf log
D: touch      100644  1 (   0,   0) 11253 /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf
D:  fsmRename (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf, /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf-RPMDELETE) 
D:  fsmUnlink (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf-RPMDELETE) 
D: file /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf vanished unexpectedly
D: create     100644  1 (   0,   0) 11253 /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf
D:   fsmOpen (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf [10]) 
D:  fsmUnpack (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf 11253 bytes [10]) 
D:  fsmChown (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf, 0, 0) 
D:  fsmChmod (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf, 00644) 
D:  fsmUtime (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf, 0x61af8a85) 
D: fsverity no signature for: path /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf dest /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf
D: lsetfilecon: (/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf, system_u:object_r:usr_t:s0) 
D: skip       100644  1 (   0,   0) 11253 /usr/share/dbus-1/system.d/org.freedesktop.machine1.conf
```

dbus notices that live and tries to parse the file I guess while rpm still works on it:
```
Jan 10 18:53:29 linux dbus-daemon[646]: dbus-daemon[646]: Encountered error 'Failed to open "/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf": Permission denied' while parsing '/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf'
Jan 10 18:53:29 linux dbus-daemon[646]: Encountered error 'Failed to open "/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf": Permission denied' while parsing '/usr/share/dbus-1/system.d/org.freedesktop.machine1.conf'
```

After that access of the affected service is blocked.

I think it's a fallout of 13f70e3 which sets fp->suffx to NULL which means fsmVerify gets called here https://github.com/rpm-software-management/rpm/blob/ae3d2d234ae47ff85229d3fce97a266fa1aa5a61/lib/fsm.c#L947-L948
Contrary to it's name `fsmVerify()` actually renames and unlinks files :-)

**Comment by pmatilai** [comment]
Right, there that test for fp->suffix should probably also test for FA_TOUCH similarly to what was there only in a different place.

Note that the whole _minimize_writes thing is problematic in more ways than just security issues and I'm not at all sure it can live on in the world where the symlink CVE cases have been fixed.

**Comment by pmatilai** [comment]
Oh, the following code (about file getting removed unexpectedly, commit 886c24cfc6c0fec90d8db1406a0e32c0e09e92c9) is equally wrong because it relies on RPMERR_ENOENT actually meaning something...

**Comment by pmatilai** [comment]
Should be fixed in commit 03f146e955e099c24f7a49eff461645e1c0a0fd7 now.
Thanks for reporting!

**Comment by lnussel** [comment]
Thanks!
